modules/host/update.go

```diff
@@ -97,23 +101,36 @@ func (h *Host) initConsensusSubscription() error {
 		// structure.
 		return h.initRescan()
 	}
-	return err
+	if err != nil {
+		return err
+	}
+	h.tg.OnStop(func() {
+		h.cs.Unsubscribe(h)
+	})
+	return nil
 }
 
 // ProcessConsensusChange will be called by the consensus set every time there
 // is a change to the blockchain.
 func (h *Host) ProcessConsensusChange(cc modules.ConsensusChange) {
-	h.mu.Lock()
-	defer h.mu.Unlock()
 	err := h.tg.Add()
 	if err != nil {
 		return
 	}
-	defer h.tg.Done()
+	wg := new(sync.WaitGroup)
+	defer func() {
+		wg.Wait()
```

Comment:
The `wg.Wait()` call needs to happen after `h.mu.Unlock()`. Because defers are called in reverse, if we are going to `defer wg.Wait()`, then that defer statement needs to come before `defer h.mu.Unlock()`, so we'd need to do manual unlocking in the rest of the function.

I will add more comments to explain this, since it is far away from the other code it is coupled to.